### PR TITLE
CM-40909 - Implement Go restore support for SCA

### PR DESCRIPTION
**Pull request by naftalicy**

### CODE REVIEW

File: cycode/cli/utils/shell_executor.py [hunk 1]
```diff
@@ -9,19 +9,26 @@
...
-            shell=execute_in_shell,
+            shell=False,
             check=True,
             capture_output=True,
+            text=True,
         )
 
+        # Write stdout output to the file if output_file_path is provided
+        if output_file_path:
+            with open(output_file_path, 'w') as output_file:
```
[the remark this answers is not in the document]
Comment (naftalicy) [reply cycode/cli/utils/shell_executor.py]: #cycode_sast_false_positive

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        return ['cd', self.prepare_tree_file_path_for_command(manifest_file_path), '&&', 'go', 'list', '-m', '-json']
+
+    def get_lock_file_name(self) -> str:
+        return GO_RESTORE_FILE_NAME
+
+    def verify_restore_file_already_exist(self, restore_file_path: str) -> bool:
+        return os.path.isfile(restore_file_path)
+
+    def prepare_tree_file_path_for_command(self, manifest_file_path: str) -> str:
+        return manifest_file_path.replace('/' + BUILD_GO_FILE_NAME, '')
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/go/restore_go_dependencies.py:31]: os.sep as in prev PR

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from cycode.cli.models import Document
+
+GO_PROJECT_FILE_EXTENSIONS = ['.mod']
+GO_RESTORE_FILE_NAME = 'go.sum'
+BUILD_GO_FILE_NAME = 'go.mod'
+
+
+class RestoreGoDependencies(BaseRestoreDependencies):
+    def __init__(self, context: click.Context, is_git_diff: bool, command_timeout: int) -> None:
+        super().__init__(context, is_git_diff, command_timeout, True)
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/go/restore_go_dependencies.py:16]: named arg instead as in prev PR

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    def is_project(self, document: Document) -> bool:
+        return any(document.path.endswith(ext) for ext in GO_PROJECT_FILE_EXTENSIONS)
+
+    def get_command(self, manifest_file_path: str) -> List[str]:
+        return ['cd', self.prepare_tree_file_path_for_command(manifest_file_path), '&&', 'go', 'list', '-m', '-json']
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/go/restore_go_dependencies.py:22]: are you sure that this command will be executed on Windows normally? `cd` and `&&` worry me

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+GO_RESTORE_FILE_NAME = 'go.sum'
+BUILD_GO_FILE_NAME = 'go.mod'
+
+
+class RestoreGoDependencies(BaseRestoreDependencies):
+    def __init__(self, context: click.Context, is_git_diff: bool, command_timeout: int) -> None:
+        super().__init__(context, is_git_diff, command_timeout, True)
+
+    def is_project(self, document: Document) -> bool:
+        return any(document.path.endswith(ext) for ext in GO_PROJECT_FILE_EXTENSIONS)
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/go/restore_go_dependencies.py:19]: `endswith` accepts a tuple with many options. to not do for loop manually

File: cycode/cli/files_collector/sca/maven/restore_maven_dependencies.py [hunk 6]
```diff
@@ -15,11 +15,12 @@
 BUILD_MAVEN_FILE_NAME = 'pom.xml'
 MAVEN_CYCLONE_DEP_TREE_FILE_NAME = 'bom.json'
 MAVEN_DEP_TREE_FILE_NAME = 'bcde.mvndeps'
+OUTPUT_FILE_MANUALLY = False
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/maven/restore_maven_dependencies.py:18]: remove as discussed in prev PR; we made false by default

File: cycode/cli/files_collector/sca/nuget/restore_nuget_dependencies.py [hunk 7]
```diff
@@ -8,11 +8,12 @@
 
 NUGET_PROJECT_FILE_EXTENSIONS = ['.csproj', '.vbproj']
 NUGET_LOCK_FILE_NAME = 'packages.lock.json'
+OUTPUT_FILE_MANUALLY = False
 
 
 class RestoreNugetDependencies(BaseRestoreDependencies):
     def __init__(self, context: click.Context, is_git_diff: bool, command_timeout: int) -> None:
-        super().__init__(context, is_git_diff, command_timeout)
+        super().__init__(context, is_git_diff, command_timeout, OUTPUT_FILE_MANUALLY)
```
Comment (MarshalX) [inline cycode/cli/files_collector/sca/nuget/restore_nuget_dependencies.py:16]: same

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py (hunk 2 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/go/restore_go_dependencies.py:31]: done

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py (hunk 3 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/go/restore_go_dependencies.py:16]: done

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py (hunk 4 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/go/restore_go_dependencies.py:22]: they works on powershell as well

File: cycode/cli/files_collector/sca/maven/restore_maven_dependencies.py (hunk 6 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/maven/restore_maven_dependencies.py:18]: done

File: cycode/cli/files_collector/sca/nuget/restore_nuget_dependencies.py (hunk 7 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/nuget/restore_nuget_dependencies.py:16]: done

File: cycode/cli/files_collector/sca/go/restore_go_dependencies.py (hunk 4 above)
Comment (naftalicy) [reply cycode/cli/files_collector/sca/go/restore_go_dependencies.py:22]: no, the CD and && works properly on cmd as well

**Review by MarshalX — APPROVED** [review]
well done!

**Pull request merged**